### DISCUSSION TITLE
Support direct upgrade from 3.9.X -> 3.12.X

### DISCUSSION THREAD

**Discussion by mitchdz** [rabbitmq/rabbitmq-server#11938] [Questions]
Hello, I'm from Canonical and working on improving the rabbitmq-server package in our archives.

You may have seen https://github.com/rabbitmq/rabbitmq-server/discussions/11418 where a user upgraded from Jammy to Noble (22.04 -> 24.04) and had issues due to feature flags. Is there any possibility in getting support for a direct upgrade from 3.9.X -> 3.12.X?

I understand the correct approach is to upgrade to the interim 3.10 -> 3.11 releases and then upgrade to 3.12, but that's not a very good solution from doing distro upgrades and can be error prone in our maintainer scripts.

Related Ubuntu bug - https://bugs.launchpad.net/ubuntu/+source/rabbitmq-server/+bug/2074309

P.S. I know our versions are older at this point, I'd like to start back porting newer versions, but would have to support a proper upgrade path to make this happen.

**Comment by lukebakken** [comment]
> Is there any possibility in getting support for a direct upgrade from 3.9.X -> 3.12.X?

No, there is not. RabbitMQ 3.9, 3.10 and 3.11 are unsupported, and RabbitMQ 3.12 is only supported for commercial customers. I don't have anything to add to @mkuratczyk's [answer](https://github.com/rabbitmq/rabbitmq-server/discussions/11418#discussioncomment-9725119), which is 100% correct.

@michaelklishin may have more to add here.

**Reply by michaelklishin** [reply]
@mitchdz so you are asking maintainers of an open source tool to support direct upgrades from version A to B, both of which are out of community support — and 3.9 has reach EOL three years ago or so — to provide you with a way to go straight from A to B just because your distribution is always at least one, often two or more, years behind.

Now that's some guts.

The answer is no, there will be no more 3.9, 3.10, 3.11 or 3.12 releases, and we will not "refactor" 3.12 to not require some feature flags that first shipped in 3.10.

Teams RabbitMQ provides our own Debian packages of both Erlang and RabbitMQ exactly because Debian and Ubuntu standard repositories have always been so slow to update, no one in our community uses the standard repo (or rather, if they do, they do it by accident).

Distributions can provide the latest package without any upgrade guarantees (since in four years between LTS releases, there almost certainly will be some feature flags required by a recent version that have been first introduced 2-3 series ago).

Or they can roughly keep up with our release pace, like Fedora does (most of the time). But no, with the Debian release cycle and Ubuntu LTS cycles of several years, I don't think that's realistic. Team RabbitMQ ships minors roughly every 6-8 months, sometimes they graduate (make mandatory) some older feature flags, so a distribution on a 2-4 year refresh cycle will always run into a similar situation.

Offering the latest version available without any upgrade guarantees is the only practical solution (that I can think of) for a distribution maintainer.

**Reply by mitchdz** [reply]
Thank you for your thorough comments Luke and Michael.

I apologies if I came off as arrogant with this request, being a package maintainer I was curious if there was an easy way to support this upgrade - I by no means expected significant work from you, but figured I would ask just in case.

It is unfortunate how far behind we (debian/ubuntu) are, and it's something I would like to help address. As you mentioned, our release cycle can make things difficult even if we keep up on it.

Again, thank you for your quick responses. I'm hoping to make this package a bit more healthy, and soon help contribute more to the upstream packaging.
